controller-server/src/main/java/com/yahoo/vespa/hosted/controller/endpointcertificates/EndpointCertificateManager.java

```diff
@@ -42,32 +51,58 @@
         this.zoneRegistry = zoneRegistry;
         this.curator = curator;
         this.secretStore = secretStore;
         this.applicationCertificateProvider = applicationCertificateProvider;
         this.clock = clock;
+        this.flagSource = flagSource;
     }
 
     public Optional<EndpointCertificateMetadata> getEndpointCertificateMetadata(Instance instance, ZoneId zone) {
 
         if (!zoneRegistry.zones().directlyRouted().ids().contains(zone)) return Optional.empty();
 
-        // Re-use certificate if already provisioned
-        Optional<EndpointCertificateMetadata> endpointCertificateMetadata =
+        // Re-use existing certificate if already provisioned
+        var endpointCertificateMetadata =
                 curator.readEndpointCertificateMetadata(instance.id())
-                        .or(() -> Optional.of(provisionEndpointCertificate(instance)));
+                        .orElse(provisionEndpointCertificate(instance));
+
+        // If feature flag set for application, look for and use refreshed certificate
+        var useRefreshedEndpointCertificate = Flags.USE_REFRESHED_ENDPOINT_CERTIFICATE.bindTo(flagSource);
+        if (useRefreshedEndpointCertificate.with(FetchVector.Dimension.APPLICATION_ID, instance.id().serializedForm()).value()) {
+            var latestAvailableVersion = greatestVersionInSecretStore(endpointCertificateMetadata);
+
+            if (latestAvailableVersion.isPresent() && latestAvailableVersion.getAsInt() > endpointCertificateMetadata.version()) {
+                var refreshedCertificateMetadata = new EndpointCertificateMetadata(
+                        endpointCertificateMetadata.keyName(),
+                        endpointCertificateMetadata.certName(),
+                        latestAvailableVersion.getAsInt()
+                );
+
+                if (verifyEndpointCertificate(refreshedCertificateMetadata, instance, zone, "Did not refresh, problems with refreshed certificate: "))
+                    return Optional.of(refreshedCertificateMetadata);
+            }
+        }
+
+        // Only log warnings
+        verifyEndpointCertificate(endpointCertificateMetadata, instance, zone, "Problems while verifying certificate: ");
+
+        return Optional.of(endpointCertificateMetadata);
+    }
 
-        // Only logs warnings for now
-        endpointCertificateMetadata.ifPresent(certificateMetadata -> verifyEndpointCertificate(certificateMetadata, instance, zone));
+    private OptionalInt greatestVersionInSecretStore(EndpointCertificateMetadata originalCertificateMetadata) {
+        var certVersions = new HashSet<>(secretStore.listSecretVersions(originalCertificateMetadata.certName()));
+        var keyVersions = new HashSet<>(secretStore.listSecretVersions(originalCertificateMetadata.keyName()));
 
-        return endpointCertificateMetadata;
+        return Sets.union(certVersions, keyVersions).stream().mapToInt(Integer::intValue).max();
```

Comment:
This will select the highest keyversion and certversion. We should only consider cert since we might be in the middle of a refresh (where keyversion might be incremented, but certversion still not ready)